### PR TITLE
Add a liveness timer to yamux_session

### PR DESCRIPTION
**Pull request by madninja**
The livenss timer gets kicked every time the session receives data. If
it ever fires it sends a ping to see if the other side is still
there. If that ping succeeds, it re-kicks the timer, but if it fails
the sesison is closed/stopped.

### CODE REVIEW

File: src/libp2p_yamux_session.erl [hunk 1]
```diff
@@ -359,6 +388,12 @@ ping_receive(#header{length=PingID}, State=#state{pings=Pings}) ->
     % with the ping time
     case maps:take(PingID, Pings) of
         error -> State;
+        {{liveness, TimerRef, _}, NewPings} ->
+            %% When we receive a ping sent from a liveness check we
+            %% jus tcancel the timer and kick the liveness timer down
+            %% the road again.
+            erlang:cancel_timer(TimerRef),
```
Comment (Vagabond) [inline src/libp2p_yamux_session.erl]: isn't the cancel here redundant given the kick also cancels it?

File: src/libp2p_yamux_session.erl [hunk 2]
```diff
@@ -359,6 +388,12 @@ ping_receive(#header{length=PingID}, State=#state{pings=Pings}) ->
     % with the ping time
     case maps:take(PingID, Pings) of
         error -> State;
+        {{liveness, TimerRef, _}, NewPings} ->
+            %% When we receive a ping sent from a liveness check we
+            %% jus tcancel the timer and kick the liveness timer down
```
Comment (Vagabond) [inline src/libp2p_yamux_session.erl]: Typo 'jus'

File: src/libp2p_yamux_session.erl [hunk 3]
```diff
@@ -164,6 +170,11 @@ handle_info({send_result, Key, Result}, State=#state{sends=Sends}) ->
 
 handle_info({timeout_ping, PingID}, State=#state{}) ->
     {noreply, ping_timeout(PingID, State)};
+handle_info(timeout_liveness, State=#state{}) ->
+    %% No data received.. Send a ping
+    {noreply, ping_send(liveness, State)};
+handle_info(liveness_failed, State=#state{}) ->
+    {stop, {error, liveness}, State};
```
Comment (Vagabond) [inline src/libp2p_yamux_session.erl]: I'd rather you log and exit normal here, so we avoid a crash report?

File: src/libp2p_yamux_session.erl [hunk 4]
```diff
@@ -341,10 +359,21 @@ handle_send_result({call, From}, Result, State=#state{}) ->
...
+ping_send(From, State=#state{next_ping_id=NextPingID}) ->
+    session_send({ping, From, NextPingID}, header_ping(NextPingID), ?TIMEOUT,
+                 State#state{next_ping_id=NextPingID + 1}).
+
 -spec ping_timeout(ping_id(), #state{}) -> #state{}.
 ping_timeout(PingID, State=#state{pings=Pings}) ->
     case maps:take(PingID, Pings) of
         error -> State;
+        {{liveness, _, _}, Pings2} ->
+            %% On a livenes ping timeout,the liveness check has
```
Comment (Vagabond) [inline src/libp2p_yamux_session.erl]: Typo 'livenes'

File: src/libp2p_yamux_session.erl (hunk 1 above)
Comment (Vagabond) [reply src/libp2p_yamux_session.erl]: oh sorry, this is another timer.

Comment (madninja) [reply src/libp2p_yamux_session.erl]: It is another timer indeed. Good eye

**Pull request merged**